File: new_app/lib/product_model.dart
class ProductModel {
  String? chargerId;
  String? chargerType;
  int? maxCurrent;
  double? price;
  List<String>? images;
  String? brand;
  String? chargingSpeed;
  List<String>? compatibility;
  String? warranty;
  String? connectorType;
  bool? installationIncluded;
  int? stockAvailability;
  String? description;
  String? sId;

  ProductModel({
    this.chargerId,
    this.chargerType,
    this.maxCurrent,
    this.price,
    this.images,
    this.brand,
    this.chargingSpeed,
    this.compatibility,
    this.warranty,
    this.connectorType,
    this.installationIncluded,
    this.stockAvailability,
    this.description,
    this.sId,
  });

  ProductModel.fromJson(Map<String, dynamic> json) {
    chargerId = json['chargerId'];
    chargerType = json['chargerType'];
    maxCurrent = int.tryParse(json['maxCurrent'].toString()) ?? 0;
    price = double.tryParse(json['price'].toString()) ?? 0.0;
    images = List<String>.from(json['images'] ?? []);
    brand = json['brand'];
    chargingSpeed = json['chargingSpeed'];
    compatibility = List<String>.from(json['compatibility'] ?? []);
    warranty = json['warranty'];
    connectorType = json['connectorType'];
    installationIncluded = json['installationIncluded'] ?? false;
    stockAvailability = int.tryParse(json['stockAvailability'].toString()) ?? 0;
    description = json['description'];
    sId = json['_id'];
  }

  Map<String, dynamic> toJson() {
    return {
      'chargerId': chargerId,
      'chargerType': chargerType,
      'maxCurrent': maxCurrent,
      'price': price,
      'images': images,
      'brand': brand,
      'chargingSpeed': chargingSpeed,
      'compatibility': compatibility,
      'warranty': warranty,
      'connectorType': connectorType,
      'installationIncluded': installationIncluded,
      'stockAvailability': stockAvailability,
      'description': description,
      '_id': sId,
    };
  }
}

File: backendEcomm/src/models/product_model.js
const { Schema, model } = require('mongoose');

const productSchema = new Schema({
    category: { type: Schema.Types.ObjectId, ref: 'Category', required: true },
    chargerId: { type: String, required: true },
    chargerType: { type: String, required: true }, 
    maxCurrent: { type: Number, required: true },
    price: { type: Number, required: true },
    images: { type: [String], default: [] },
    brand: { type: String, required: true },
    chargingSpeed: { type: String, required: true },
    compatibility: { type: [String], required: true }, 
    warranty: { type: String, required: true },
    connector: {
        type: {
            type: String, 
            enum: ["single", "dual"], 
            required: true
        },
        details: { type: String, default: "" } 
    },
    installationIncluded: { type: Boolean, required: true },
    stockAvailability: { type: Number, required: true },
    description: { type: String, default: "" },
    averageRating: { type: Number, default: 0 }, // Add average rating field
    updatedOn: { type: Date },
    createdOn: { type: Date }
});

// Middleware to handle timestamps
productSchema.pre('save', function (next) {
    this.updatedOn = new Date();
    this.createdOn = new Date();
    next();
});

productSchema.pre(['update', 'findOneAndUpdate', 'updateOne'], function (next) {
    const update = this.getUpdate();
    delete update._id;
    update.updatedOn = new Date();
    next();
});

const ProductModel = model('Product', productSchema);

module.exports = ProductModel;
